Fixed Timestep Interpolation (3D) (3.x)

lawnjelly opened the pull request:
Adds fixed timestep interpolation to the visual server.
Switchable on and off with project setting.

On request from reduz I have changed the `teleport` function name to `reset_physics_interpolation`.

(While some of the features will be shared with 2D, they are both fairly separable so I will be addressing 2D in a later PR. Once we are happy with the approach in 3.x, I will do similar for 4.x, perhaps after we get some initial testing.)

## Details
* Interpolation lists in the `VisualServer` are now stored in the `Scenario` (i.e. unique per scene tree).
* There is an overall on / off bool setting per scene tree which can be changed by calling `set_scene_tree_physics_interpolation_enabled()` on the `SceneTree`, which defaults to the project setting `physics/common/physics_interpolation` (and is always off when `is_editor_hint` is set).
* All nodes have a `physics_interpolated` flag, which can be read and set for individual control with `set_physics_interpolated()`.
* There is now a mechanism for retrieving interpolated transforms on frames : The command `get_global_transform_interpolated()`.
* `CPUParticles` and `MultiMeshInstances` now work via a modified `MultiMesh` that can handle interpolation.

Based around proposal : https://github.com/godotengine/godot-proposals/issues/2753

## Notes
* This version often requires very little changes in terms of game code, all that is required is making sure that movement of objects and camera take place in `_physics_process`. Where changes are necessary these are most likely to occur in custom `Camera` code.
* I moved the 'global' on / off setting, from `Engine` to `SceneTree` in response to comments from reduz and jordo. The interpolation lists are also moved to be per `Scenario` (i.e. SceneTree). This enables the whole system to be flexible for users who are not using the standard SceneTree.
* As a consequence of moving the on / off setting to the `Scenario`, the `Camera` needs quick access to its scenario when calling `camera_set_transform`. As a result I've added storing the `scenario` in the `Camera`, in a similar manner to how it is stored in `Instance`s.
* I've moved the interpolation function for the `Transform` into a separate class `TransformInterpolator`. This was getting repeated in several places in the code (for instances, cameras, and potentially in the scene tree code) and it probably makes sense to put it in one place.
* Having the interpolation code in one place makes it easier to adjust it, if we want to add e.g. Hermite in future.
* The interpolation code automatically switches to a `slerp` where possible (this is dependent on the `Basis`).
* We may want to eventually address audio as well, although I'm not sure it needs to be done in the same PR.
* I removed changes to `InterpolatedCamera`. These can be done in a separate PR once we decide on an approach.
* There is a checksum optimization for checking for noop `set_transform`s. This isn't strictly speaking necessary, so try not to get confused by this (it's cheaper than comparing 2 transforms each time, which could be important with a bunch of non-moving objects).

## get_global_transform_interpolated()
It turns out that in a practical sense, especially for Cameras, it is often necessary to turn off the automatic fixed timestep interpolation and interpolate manually. See below for the problems with @Calinou 's project for an example - mouse looks better without fixed timestep interpolation, either with no interpolation, or using non-fixed timestep based interpolation. This means in `_process` users will want to be able to focus on the _displayed_ (i.e. interpolated) position of an object, rather than the position at the last physics tick.
The same issue of needing to know the transform on a frame rather than a physics tick occurs when you need to make something emit at `_process` rather than `_physics_process`.
For these reasons, the PR includes an implementation for selectively duplicating the interpolation client side, i.e. within the scene tree. The reason for this requirement is that interpolation usually takes place in the VisualServer, and the queue / threading system means that synchronization is difficult to retrieve information from the VisualServer - it may cause a stall.
The solution used involves storing extra optional `InterpolationData` on the `Spatial`. This mechanism only is activated the first time the user calls `get_global_transform_interpolated()` on a node. This means there is no cost for the vast majority of nodes, both in terms of memory and processing.

## MultiMeshes
It turned out that a large hurdle was handling MultiMeshes in a reasonable manner. As well as for MultiMeshInstances, these are used by CPUParticles under the hood. MultiMeshes use instancing in GLES3.

As a multimesh is a list of instances, in theory the main work is to maintain a list of previous and current data (transforms, colors and custom data), and interpolate between this. In practice however there are two snags:

1) From CPUParticles, the order of the instances (particles) is not coherent from frame to frame. They may be drawn in e.g. z draw order etc. This means we cannot use a simple round robin list of the previous and current data for each instance.

The solution I have used here is the simplest conceptually, but it does impose some overheads. Basically for each bulk update of the particle instances (on a physics tick), instead of sending just the current data, it sends the current AND previous data every tick. This is wasteful, but it ensures that the data of the current and previous is always in sync.

This is achieved via a new alternative function:
```
virtual void multimesh_set_as_bulk_array_interpolated(RID p_multimesh, const PoolVector<float> &p_array, const PoolVector<float> &p_array_prev);
```
2) In normal use by users it is expected the instances are more likely to remain in sync and the usual:
```
multimesh_instance_set_transform
multimesh_instance_set_color
multimesh_instance_set_custom_data
```
functions can be used as usual. There is also one other new function of interest:
```
virtual void multimesh_instance_reset_physics_interpolation(RID p_multimesh, int p_index);
```
which can be used to teleport individual instances.

## How MultiMesh works internally
MultiMesh was previously implemented purely as classes within each backend, which was a problem, because ideally we want to share interpolation functionality between the two.
The solution was two create a set of wrapper functions in `RasterizerStorage`, which handle interpolation, and in turn call a similar set of underscored functions in the actual backends which are identical to the previous code.
In order to store interpolation specific data, the backends are responsible for creating a `MMInterpolator` (multimesh interpolator) object as part of their multimesh, and returning it when the following function is called:
```
virtual MMInterpolator *_multimesh_get_interpolator(RID p_multimesh) const = 0;
```
From here the Rasterizer storage can now handle storing previous and current data for each MultiMesh, and interpolating before a frame is drawn in a similar way to the Instances and Cameras.

One further gotcha is that `MultiMesh` does not have a `Scenario` associated with it, and could potentially be used like a resource in several MultiMeshInstances, potentially in different `Scenario`s (although unlikely). As the global interpolation flag is now a property of the `SceneTree / Scenario`, this makes the situation not quite so ideal, but a flag for `interpolated` is available on each `MultiMesh` and will be set by the corresponding instance that "claims" it.

This isn't perfect but it's a bodge around the problem that a `MultiMesh` isn't owned by a `Scenario`. In the vast majority of cases anyway the global interpolation will either be on or off, and providing a `MultiMesh` isn't used in two Scenarios with different `SceneTree` interpolation flags this shouldn't be a problem.

## CPUParticles (in global (non-local) coordinate mode)
Along with the multimesh changes, there were additional changes to CPU particles to get them to behave correctly with interpolation.

### Particle global coordinate space
CPUParticles were previously implemented in a slightly curious way in global mode. Although the particle positions are calculated in global (world) space, instead of specifying the particles in global coordinates, the instance in the VisualServer still had the transform inherited from the parent node. In order to compensate for this the particles are given an extra step to _back transform_ them from global space to local space, by xforming by the affine inverse of the node transform.
This was seen as potentially problematic in terms of interpolation, and also was requiring more CPU than necessary. An obvious simpler solution is to set the instance to an identity transform, and just specify the particles directly in global space.
I have done this in the PR, and it did require some small adjustments but seems to be working correctly. It is possible that there may be more small compensations I have missed, but this seems well worth doing.

## Camera code
The most likely area to require changes for existing games is Camera code. While cameras _can_ be used as before except updating only in `_physics_process` and allowing fixed timestep interpolation to interpolate them, best results often involve turning automatic interpolation off and doing this manually.
When doing manual interpolation, as with the _smoothing addon_, it is recommended that in the SceneTree you separate the Camera from the object it is targetting, i.e. it is easier if the Camera is specified in global space and is not a child / grandchild of the object being targetted.
For custom interpolation, I recommend moving the Camera in `_process`, and calling `get_global_transform_interpolated()` on the target node, and using this interpolated position to focus on. The other calculations can be done as would normally be done for a camera.

## Testing
Tested so far in:
* Truck town demo
* 3d platformer demo
* Blubits demo
* Test framework
* SuperTuxParty
* Veraball (some changes to Camera gdscript, see later in this thread)
* Wrought Flesh

Works great in all so far at e.g. 20 ticks per second. In most cases no changes were needed to the games, just switching on interpolation.

<!--
Pull requests should always be made for the `master` branch first, as that's
where development happens and the source of all future stable release branches.

Relevant fixes are cherry-picked for stable branches as needed.

Do not create a pull request for stable branches unless the change is already
available in the `master` branch and it cannot be easily cherry-picked.
Alternatively, if the change is only relevant for that branch (e.g. rendering
fixes for the 3.2 branch).
-->

lawnjelly committed:
Fixed Timestep Interpolation (3D)

Adds fixed timestep interpolation to the visual server.
Switchable on and off with project setting.

This version does not add new API for set_transform etc, when nodes have the interpolated flag set they will always use interpolation.

lawnjelly commented:
This should be getting pretty reasonable now, I could do with some people testing this out. The interpolation defaults to on (`project_settings/physics/common/physics_interpolation`), although it will likely default to off later for backward compatibility.

## Instructions
* Download a build from the green tick section.
* Make sure the game you are testing moves objects and cameras in `_physics_process` instead of `_process`.
* Try changing the physics tick rate to e.g. 20 tps (`physics/common/physics_fps`).

Let me know any bugs / strange behaviour. Strictly speaking you'll also want to call `teleport` in some situations, but that's more for making releases.

All nodes have a `physics_interpolated` property (which you can either set in the editor, or with `set_physics_interpolated()`). These default to true. The property is recursive within branches, if you set to true or false, all children will also be altered.

Calinou commented:
I modified a game I made a while ago to use built-in physics interpolation instead of the smoothing add-on: https://github.com/Veraball/veraball/tree/builtin-physics-interpolation-pr
Physics interpolation is enabled by default. Press <kbd>P</kbd> to toggle it at run-time.

It works well overall, but I noticed a few issues:

- Particles with **Local Coords** disabled parented to a physics body will move around in an erratic manner.
- Objects near the camera will blink rapidly if they move at a high speed when using low physics FPS. This is most likely difficult to notice at a physics FPS of 60, but it can happen more often with physics FPS set to 30 or less.

To make differences more obvious, the physics FPS is set to 10 in both videos.

### Without physics interpolation

https://user-images.githubusercontent.com/180032/142765104-1c6fd553-473a-4f1e-9186-d5f20432e2db.mp4

### With physics interpolation

***Epilepsy warning:** The video below contains rapid flashes.*

https://user-images.githubusercontent.com/180032/142765108-529c920e-42da-4ce9-931f-e6838c890562.mp4

lawnjelly commented:
Great feedback thanks @Calinou ! :grin: 

Yeah the particles will be happening thinking about it because there is no interpolation on them yet, I didn't go through all the objects at the early stage because there were 4 different PRs and we needed to reach agreement on various things before I finished off. But now reduz is happy with the methods etc, I should be able to fix these.

At the moment the interpolation is 3D only (only doing 3D in this PR, but much of it should be shared when I do 2D), and it's only acting on Cameras and Instances so far. This covers most of the 3D objects but there are exceptions like particle systems, but I'm expecting these to be quite straight forward.

The frustum culling is probably currently based on the most recent transform, rather than the interpolated value, I should be able to fix this. :+1:

Calinou commented:
The latest commit fixes frustum culling and particle jittering (at the cost of particles leading the ball when moving). Thanks :slightly_smiling_face: 

In my project, I noticed that camera rotation jitters when physics interpolation is enabled:

https://user-images.githubusercontent.com/180032/142887906-32d8bc19-2d9a-4e43-9613-4aa379519f36.mp4

This doesn't occur when physics interpolation is disabled:

https://user-images.githubusercontent.com/180032/142887829-ff8206ff-2e8e-4b8d-a294-83e319805ab0.mp4

The way I do camera rotation in my project is quite hacky (it involves two nodes with one of them acting as a Y pivot), so this may be the cause.

lawnjelly commented:
Yeah cameras are a special case, in a lot of situations I actually recommend turning off interpolation on them with `set_physics_interpolated(false)` and doing the interpolation yourself (I've pretty much ended up doing this in most interpolated games). You may be able to get away with it at e.g. 60 ticks per second, but it becomes more of a problem at lower tick rates.

With a mouse look you probably want to use interpolation for the target position, but not for the mouse look. This may not even be that sensible to try VisualServer side, because like the situation with `InterpolatedCamera`, there are so many variations possible, with combinations of mouse look and automated camera etc.

This does mean however if you do this manually that for following an interpolated target, you will have to calculate the interpolation for the target yourself manually (as detailed in my post above).

It could also be that in your situation there is some interaction with your pivots.

lawnjelly commented:
@Calinou the new version is now quite improved. To solve your camera issue with mouse look I made the following changes to your project:
Added this line to ready() for the ball.gd, to turn off interpolation for the camera branch:
```
func _ready() -> void:
	$Yaw.set_physics_interpolated(false)
```
And moved the update of the camera etc from the `_physics_process` to `_process` (so it updates each frame), and changed it to call the new `get_global_transform_interpolated()` function.
```
func _process(delta):
	var pos = body_node.get_global_transform_interpolated().origin
	
	# Move camera, sample players and ray with the ball (but don't rotate them)
	if Game.camera_follows_ball:
		yaw_node.set_translation(pos)
	ray_node.set_translation(pos)
	sounds_node.set_translation(pos)
	night_light_node.set_translation(pos)
	boost_light_node.set_translation(pos)
```
Once this was done I'm getting a nice smooth camera, and the particles now are no longer leading the ball thanks to the other fixes in the PR.

lawnjelly commented:
I've been testing out the new interpolated `CPUParticles` and they are pretty awesome, and should run faster than before thanks to not needing the reverse transform applying

This is 2 physics ticks per second, with a box following a sine wave, and global particles following the box.

Without interpolation:
https://user-images.githubusercontent.com/21999379/143242198-2ac8d80f-ee9d-4cd5-ade4-1fd753e102c1.mp4

With interpolation:
https://user-images.githubusercontent.com/21999379/143242308-2a456cdf-10b3-477c-9ddb-5cd4a640a876.mp4

Calinou commented:
> I've been testing out the new interpolated CPUParticles and they are pretty awesome, and should run faster than before thanks to not needing the reverse transform applying

Good to hear! This will also resolve https://github.com/godotengine/godot/issues/47973 then :slightly_smiling_face:

lawnjelly commented:
> > I've been testing out the new interpolated CPUParticles and they are pretty awesome, and should run faster than before thanks to not needing the reverse transform applying
> 
> Good to hear! This may help resolve #47973 then slightly_smiling_face

It won't resolve that issue completely, as it seems to be that is about the particles being emitted on a frame basis. They still are here, but if you are running at e.g. 60fps, and the physics is 10tps, you'll get a spread over 60 positions instead of just 10.

That issue could be solved though with the fixed timestep interpolation data, at least for CPUParticles. It's not super tricky, as the poster said in that issue it's just a matter of predicting where the emitter would be at the emission time, but it's probably best left to a separate PR. And it would be compatibility breaking unless we added a separate mode for this.. i.e. it will look slightly different. But then again I guess people will be expecting fixed timestep interpolation to look different anyway, so we could fix it as part of the same group of changes.

For GPUParticles I'm not really familiar with how they operate so no idea on that front. :thinking:

lawnjelly commented:
## Slerping and special cases

### Characters
Everything seems to be working pretty good now, the one remaining thing I've noticed is that in some games characters are slerping less than would be expected. This isn't necessarily a problem, but it does mean that really low tick rates (less than 10tps) are not advisable in affected games.

To show what will happen, this is the TPS demo character running at 5tps or less:
![warping](https://user-images.githubusercontent.com/21999379/143553575-9f03a581-3e84-48ce-a422-05348d1e4415.png)

### Basis lerping
You can see that there is warping going on, because it has decided it cannot slerp, so is reverting to basis lerping. Basis lerping is quick and has the advantage it works with non-standard rotations (shears etc), but it will introduce some warping depending on how low the tick rate and how fast the objects are rotating.

### VisualServer versus smoothing addon
I was wondering why the smoothing addon is more reliable at producing slerping - and after a little investigation it seems this is a consequence of the paradigm of doing the interpolation in the VisualServer.

The VisualServer only has access to the global space Basis of e.g. the `VisualInstance`. What is happening is that if you create a character with a Spatial or RigidBody on top and the skeleton etc as children, that the actual yaw is applied to the Spatial as a simple correct rotation basis. But the basis sent to the visual server is _after_ this has been concatenated into the mesh transform. This can often end up not being a valid rotation basis, hence lerping is used.

The smoothing addon in contrast is applied in the SceneTree directly to the root Spatial of the character, thus slerping is easily possible.

We did bring this up as a possibility when discussing the proposal, the example used was a house on the surface of a rotating parent planet. That if the interpolation was done at the level of the planet, the house would follow the surface of the planet exactly, but if done at the level of the mesh, then the house would interpenetrate the planet surface in between physics ticks. The end decision was that if it was run at a high enough tick rate, this would not be noticeable, and the simplicity advantages of running in the VisualServer outweighed this.

So it turns out the same effect is happening with the characters, so that will put a lower limit on the tick rate you can get away with, but I guess most people will be running above say 20 ticks per second. The physics generally won't work well at super low tick rates anyway.

And if necessary it is possible to override the automatic interpolation for characters using `set_physics_interpolated()` on the node and doing it manually, so I guess all cases are covered.

### Wheels and fast rotating objects
Another situation where it may not be possible to use a super low tick rate is when wheels or similar fast rotating objects are used in the game. The reason is that the interpolation can only interpolate between the two rotations between 2 physics ticks. If this rotation is up to 180 degrees, this can potentially work ok. But the moment you have an object rotating more than 180 degrees in a physics tick (e.g. a fast rotating wheel), you have an interesting problem, the interpolation can make the wheel interpolate the _wrong way_, i.e. it will choose the shortest arc.

This could produce some odd effects when a wheel changes between these frequencies, with phasing and backwards / forwards rotation. Indeed if the wheel rotates 360 degrees in a physics tick, it will appear to be stationary with interpolation. :grin: .

In theory you might be able to get around this by again turning off automatic interpolation and doing it manually in the scene tree side. The reason VisualServer can't cope with this is that rotations as a basis _wrap around_, whereas if you have the code that decides the wheel rotation, you may be able to interpolate by the unwrapped rotation, which would be the correct way of addressing this problem.

## Conclusion
Anyway these are just some pinch points worth mentioning in advance. In practice I suspect users will just choose a reasonable tick rate for such games, or apply workarounds.

lawnjelly commented:
It's working pretty good now. Two remaining anomalies as explained above:

## Phasing
This is running at 2 tps, with the rotation rate varying between just under 180 degrees per tick, and just over 180 degrees. As you can see, if it rotates less than 180 degrees, it interpolates one way, and more than 180 degrees, it interpolates in the incorrect direction (shortest arc).

https://user-images.githubusercontent.com/21999379/143674568-a9452b52-c9cb-4c17-8590-908432438781.mp4

## Throbbing
This is the same project above, but the box now has a non-uniform scale applied (1.1, 1, 1). This means that slerping will no longer be used and it reverts to basis lerping. As can be seen, basis lerping can cause severe warping at low tick rates. Generally it is a good idea to stay above around 20tps when lerping is being used in these cases.

https://user-images.githubusercontent.com/21999379/143674579-ad27c2db-d347-45a8-a558-10696d4c8c14.mp4

lawnjelly commented:
Moving this out of draft now as it could do with some more eyes on and testing, and identifying if any other nodes could be problematic (we don't necessarily need to cover *all* specialist nodes to start with in the first PR, but it would be nice to know if there are any others that need adjustment like the CPUParticles did).

Calinou commented:
I tested this PR again and it seems to work as expected :slightly_smiling_face: 

One question though: Should we expose `physics_interpolated` as a property in Spatial (most likely in the Transform section)? This way, you can disable it without having to do it via a script's `_ready()` function.

lawnjelly commented:
> I tested this PR again and it seems to work as expected slightly_smiling_face
> 
> One question though: Should we expose `physics_interpolated` as a property in Spatial (most likely in the Transform section)? This way, you can disable physics interpolation on a node without having to do it via a script's `_ready()` function.

I originally _did_ expose it as a property, but during discussions with reduz he wasn't altogether clear on why this was needed, given that it is mostly a global on/off switch (although bear in mind I don't think he has used fixed timestep interpolation before, the usage patterns are clearer to us simply because we have used the smoothing addon etc before). The `veraball` project however is a good example showing why the ability to turn off interpolation and do it manually in some circumstances  _is required_, hence the functions `set_physics_interpolated()` and `get_physics_interpolated()` are bound.

Personally I would also be tempted to make the `interpolated` variable editable in the inspector as a property, as in most cases you will either want to set interpolation on or off for a particular node (or branch), rather than toggle it at runtime. The binding is actually already there in the PR (line 3000, node.cpp) but commented out (_it is in Node BTW rather than Spatial because once 2D is rolled out, the functionality for turning on and off interpolation in nodes can be shared_).

On the other hand, it isn't absolutely essential we bind it when we first merge, we can easily bind it in a further PR if necessary, and it is solvable with 1 line of script to be fair.

Given the complexity of the system it may end up just you and me that review it initially, but once merged other contributors will quickly get a feel for it and be able to voice opinions for tweaks.

lawnjelly commented:
## Defaulting to on or off?
One more thing to double check is whether we want this to default to on. It does default to on at the moment but is easy to flip. reduz from memory did suggest it default to on.

I can see an argument for this in terms of getting users creating new projects with it by default from the get go (as it is easier if you use it from the start for a new project).

On the other hand I do have some worries about backward compatibility with some existing games which may do things in a way that aren't ideal for interpolation. Also consider that at the moment this default will be the same for 3D and 2D, when 2D is rolled out (given that functionality is shared, and by nature it is unlikely anyone would want interpolation in 3D but not in 2D in the same game, but not impossible I guess with a 2D minigame, but this may be more hassle than it is worth). 

## Pros
1) For creating a new project in 3D this makes sense as most games will be better with fixed timestep interpolation (FTI). Some exceptions (there may be more):

* Multiplayer games doing their own interpolation
* Fast twitch shooters wishing to do their own interpolation / avoid the 1 tick lag.
* Games doing their own interpolation to use e.g. hermite.

Having it default to on means that newbie users won't have to immediately make this choice, it will be made for them, and will be the correct choice in 99% of cases.

2) A lot of existing games running at e.g. 60tps will run fine with FTI with no changes, and get the benefits of smoothing on higher refresh rate monitors etc.

## Cons
1) FTI relies on users putting the logic in `_physics_process` rather than `_process`. There is a danger that some existing games using `_process` will look right in the IDE if the users monitor is refresh rate 60, but won't run ideally on end user machines.
2) FTI isn't necessarily the best approach for pixel based 2D games, which a lot of our users make. Snapping behaviour is quite crucial in these games, and FTI will affect this. 
3) Existing games that do their own interpolation will either want to switch FTI off in the project setting, or change their project to use the core version instead of their own. This isn't a big con, as anyone having implemented their own interpolation will understand the concepts involved and have no problem e.g. switching the core system off.

Overall as usual I am cautious and would tend towards defaulting to off and having it as an opt-in feature, but I do understand that especially in a beta it might be better to default to on to get people testing it. :thinking:

lawnjelly commented:
I've started writing some preliminary docs here, this may be useful in terms of understanding the PR and concepts involved:
https://github.com/lawnjelly/Misc/blob/master/FTIDocs/FTI.md

Oyusi commented:
I see you've gone with the global on/off route. Will it still be possible to enable/disable individual nodes with ease? I'm working on a multiplayer game where I perform hermite interp on puppets for certain things, but others I would still like to use this for local or authority nodes. I've used 'Engine.get_physics_interpolation_fraction()' in other projects, like your Smoothing addon, but a more automated solution like this PR is very welcome. Thanks.

lawnjelly commented:
> Will it still be possible to enable/disable individual nodes with ease?

Yup sure! As you say, there is control over Nodes and branches. We may also end up exposing the `interpolated` flag as a property in the inspector.

I'm aware some people might already want to use hermite, or other custom solutions for multiplayer, so you can use it in combination too, e.g. some Nodes using the built in interpolation and some using custom hermite interpolation.

The difficulty with hermite as I understand it, is that to do properly, we need the physics velocity of the objects, which aren't currently available in the VisualServer. We can derive a velocity from the previous and current position, but this is not the same thing. I'll nevertheless probably have a try with this just to see what it looks like, and we also may be able to offer an option to use extrapolation instead of interpolation. But these don't have to be present in the inital PR.

At some point I was intending to evaluate whether there is a sensible way we can pass velocities from the physics to the VisualServer to allow hermite interpolation. It may not be hugely involved, I've already toyed with the idea of providing an alternative mechanism of passing transforms to the VisualServer that is more efficient. But I can't make any promises on this front, it's more an investigation and I won't be making decisions on this alone.

Of course for multiplayer you'd probably still have to do some custom interpolation because of client / server based timing issues etc. I wrote a little note about this earlier in the preliminary docs in fact.

There's a very faint possibility also that at some point in the future we might be able to rig it up so the user can have more control over the ticks / interpolation fraction for multiplayer (say if you were sending snapshots that couldn't be guaranteed to arrive on time) but that is a way in the future.

lawnjelly commented:
Just to note here, @fire was asking for help with basis exploding problems in the IK in Godot 4, and having done a test with replacing the existing slerping with the robust slerping from this PR (i.e. the transform_interpolator class), it seems to solve the problems he was having and seems keen to make this available to other users.

The robust slerping featured here is generally quite useful (is not limited to just fixed timestep interpolation) as it can deal with scaled Bases, and will fallback to Basis lerping in the case of shears, and the current Godot slerping functionality is probably prone to exploding in cases of animation, animation importing and IK.

My view is that while we should aim to have robust slerping available for everyone to use, but it may be worth keeping the solutions local temporarily (following the Godot coding guidelines) at least until we are able to test / generalize the routines, so we are sure on an API before making them generally available directly via e.g. `Basis`, or `interpolate_with`. i.e. Get both the fixed timestep interpolation and IK merged, and tested in some betas, and if at that point they are using the same functionality we can discuss making them available and bound from the core classes (e.g. Basis).

diddykonga commented:
Love the feature as is suggested, and might be the lack of sleep asking, but I was wondering in what use cases would one want to use the "static" / "real" global transform of an interpolated node? If a node, is updated/used via _process wouldn't they always want access to the "interpolated" global transform of an interpolated node? with potentially the "static" / "real" global transform being the edge case.
Again forgive me If there are obvious use cases, where this doesn't make sense.

lawnjelly commented:
> Love the feature as is suggested, and might be the lack of sleep asking, but I was wondering in what use cases would one want to use the "static" / "target" global transform of an interpolated node? If a node, is updated/used via _process wouldn't they always want access to the "interpolated" / "main" global transform of an interpolated node? with potentially the "static" / "target" global transform being the edge case.
> 
> Again forgive me If there are obvious use cases, where this doesn't make sense.

You are absolutely right, in that calling from `_process` with interpolation switched on you would _normally_ be interested in the interpolated transform when you are using physics interpolation.

However on the flip side, using such implicit shortcuts is a major source of user confusion. While it may be obvious to you now, another user could say in future, well why doesn't `get_global_transform()` give me the global transform (non interpolated), and they would be just as correct. Or there may turn out to be valid use cases in which you may want access to the non-interpolated transform from `_process`. So we don't want to prematurely shut out this possibility.

When we discussed this, reduz expressed a preference for favouring explicit functions rather than banditing existing function calls, and I have to say I agree in this case.

In some ways the opposite is true. We may decide it is better to flag a WARN_PRINT_ONCE if users calling `get_global_transform` from `_process` in this situation. This would guard against users switching on physics_interpolation but not realising they have to make changes to their project. This may in fact be a better way of addressing this decision - make it clear to users through a warning. But I don't think there's a need to add that right away.

In practice in most cases there should only be one or two places in your code where you need to use `get_global_transform_interpolated()`, so the argument that it makes things "easier" does not seem very strong. If you need to change lots of pieces of code to use the interpolated function, you are probably doing it wrong.

lawnjelly commented:
After discussions with reduz, I've now changed the `CPUParticles` considerably and moved the interpolation to the `MultiMesh` in the VisualServer (instead of doing it client side). In addition I've modified things to allow interpolation using `MultiMeshInstance`.

This does mean that particles are no longer emitted on "in between" frames (that would have to be fixed in a separate PR, it is not strictly related to the fixed timestep interpolation).

lawnjelly commented:
Due to the change mentioned in the my comment above, interpolated `CPUParticles` are now only  emitted on a physics tick rather than a per frame basis. This can result in some "clumping" of particles at low tick rates.

This is equivalent to this issue: #47973 , except more pronounced.

For this reason I've implemented the first solution, and am wondering about offering the second:

* Add support to interpolate the start positions of particles within ticks (I've called this "streaking" to differentiate the terminology)
* Allow an optional flag to interpolate the `CPUParticleSystem` on a per frame basis scene side, rather than utilising the `MultiMesh` interpolation, as before the change (this gives smoother particles at a slight cost in lower efficiency in such situations). Very little code is needed to allow this option.

lawnjelly commented:
This is ready for review again BTW, I'm back working on the OccluderMeshes until then. :+1:

Norrox commented:
I tossed in my code in _physics_process and pressed run and I noticed some jitter, thinking for myself, this is not working?
Went back here to check the post and saw that I had to enable physics interpolation, did that and now it works great :)

Calinou commented on doc/classes/MultiMesh.xml:
```diff
@@ -136,5 +158,11 @@
 		<constant name="CUSTOM_DATA_FLOAT" value="2" enum="CustomDataFormat">
 			The [Color] passed into [method set_instance_custom_data] will use 4 floats. Use this for highest precision.
 		</constant>
+		<constant name="INTERP_QUALITY_FAST" value="0" enum="PhysicsInterpolationQuality">
```
I think it should be `INTERP_QUALITY_LOW` to be consistent with other quality enums in Godot. We use a `low/high` dichotomy as opposed to `fast/high`.

lawnjelly replied:
I'm not absolutely sure this would give the correct impression. In the vast majority of cases fast _isn't_ low quality, it is just faster. If an object is not rotating at all, fast produces the exact same result as high quality, and unless an object is rotating at high angular velocity, the quality of fast is not significantly worse than high quality.

If we call something "low quality", people are more likely to assume that they need to use "high quality" for the best result, whereas this is not the case here. "fast" is purposefully vague on this, and the tooltip points out that "high quality" is advantageous only in some specific situations.

Calinou commented on doc/classes/MultiMesh.xml:
```diff
@@ -136,5 +158,11 @@
+			Always interpolate using Basis lerping, which can produce warping artifacts in some situations.
+		</constant>
+		<constant name="INTERP_QUALITY_HIGH" value="1" enum="PhysicsInterpolationQuality">
+			Attempt to interpolate using Basis slerping where possible, otherwise fallback to lerping.
```
```suggestion
			Attempt to interpolate using Basis slerping (spherical linear interpolation) where possible, which looks smoother. If slerping is not available, fall back to lerping.
```

Calinou commented on doc/classes/MultiMesh.xml:
```diff
@@ -104,6 +122,10 @@
 		</member>
+		<member name="physics_interpolation_quality" type="int" setter="set_physics_interpolation_quality" getter="get_physics_interpolation_quality" enum="MultiMesh.PhysicsInterpolationQuality" default="0">
+			Choose whether to use an interpolation method that favors speed or quality.
+			When using low physics tick rates and / or high rates of object rotation you may get better results from the high quality setting.
```
"Low" should be defined for clarity:

```suggestion
			When using low physics tick rates (typically below 20) or high rates of object rotation, you may get better results from the high quality setting.
```

Calinou commented on doc/classes/MultiMesh.xml:
```diff
@@ -48,6 +48,14 @@
+			<return type="void" />
+			<argument index="0" name="instance" type="int" />
+			<description>
+				When using [code]physics interpolation[/code], this function allows you to prevent interpolation on an instance in the current physics tick.
```
```suggestion
				When using [i]physics interpolation[/i], this function allows you to prevent interpolation on an instance in the current physics tick.
```

Calinou commented on doc/classes/MultiMesh.xml:
```diff
@@ -57,6 +65,16 @@
+			<argument index="0" name="array_current" type="PoolRealArray" />
+			<argument index="1" name="array_previous" type="PoolRealArray" />
+			<description>
+				An alternative version of [method MultiMesh.set_as_bulk_array] which can be used with [code]physics interpolation[/code]. This method takes two arrays, and can set the data for the current and previous tick in one go. The renderer will automatically interpolate the data at each frame.
+				This is useful for situations where the order of instances may change from physics tick to tick, such as particle systems.
+				When the order of instances is coherent, the simpler [method MultiMesh.set_as_bulk_array] can still be used with interpolation.
```
```suggestion
				An alternative version of [method MultiMesh.set_as_bulk_array] which can be used with [i]physics interpolation[/i]. This method takes two arrays, and can set the data for the current and previous tick in one go. The renderer will automatically interpolate the data at each frame.
				This is useful for situations where the order of instances may change from physics tick to tick, such as particle systems.
				When the order of instances is coherent, the simpler [method MultiMesh.set_as_bulk_array] can still be used with interpolation.
```

Calinou commented on doc/classes/Spatial.xml:
```diff
@@ -19,6 +19,13 @@
+			<return type="Transform" />
+			<description>
+				When using physics interpolation, there will be circumstances in which you want to know the interpolated (displayed) transform of a node rather than the standard transform (which may only be accurate to the most recent physics tick).
+				This is particularly important for frame based operations that take place in [code]_process[/code], rather than [code]_physics_process[/code]. Examples include [Camera]s focussing on a node, or finding where to fire lasers from on a frame rather than physics tick.
```
```suggestion
				This is particularly important for frame-based operations that take place in [method Node._process], rather than [method Node._physics_process]. Examples include [Camera]s focusing on a node, or finding where to fire lasers from on a frame rather than physics tick.
```

lawnjelly commented on doc/classes/Node.xml:
```diff
@@ -627,6 +651,14 @@
 				Sets the node's network master to the peer with the given peer ID. The network master is the peer that has authority over the node on the network. Useful in conjunction with the [code]master[/code] and [code]puppet[/code] keywords. Inherited from the parent node by default, which ultimately defaults to peer ID 1 (the server). If [code]recursive[/code], the given peer is recursively set as the ...
 			</description>
 		</method>
+		<method name="set_physics_interpolated">
+			<return type="void" />
+			<argument index="0" name="p_interpolated" type="bool" />
+			<description>
+				Enables or disables physics interpolation per node, offering a finer grain of control than turning physics interpolation on and off globally.
+				[b]Note:[/b] This can be especially useful for [Camera]s, where custom interpolation can sometimes give superior results.
+			</description>
+		</method>
```
I did in fact expose it in earlier versions - when we discussed this @reduz preferred to have this hidden (I think with the aim of not overwhelming users with new concepts, or exposing the minimum possible).

On the other hand, if we get feedback expressing an interest to expose this I think we can easily re-add it.

lawnjelly commented on drivers/gles3/rasterizer_storage_gles3.cpp:
```diff
@@ -4911,6 +4911,16 @@ AABB RasterizerStorageGLES3::multimesh_get_aabb(RID p_multimesh) const {
 
+RasterizerStorage::MMInterpolator *RasterizerStorageGLES3::_multimesh_get_interpolator(RID p_multimesh) const {
+	MultiMesh *multimesh = multimesh_owner.getornull(p_multimesh);
+	// if (!multimesh) {
+	// print_line("multimesh not found " + itos(p_multimesh.get_id()));
+	// }
+	ERR_FAIL_COND_V(!multimesh, nullptr);
```
Ah yes I'll do that. :+1:

lawnjelly commented on scene/main/node.h:
```diff
@@ -120,25 +117,47 @@ class Node : public Object {
+		int process_priority;
+
 		// variables used to properly sort the node when processing, ignored otherwise
 		//should move all the stuff below to bits
```
It's to group the bools together. :slightly_smiling_face: 
As they are now 1 bit each, it's important to define them together to get the benefit so they are stored in one continuous `uint`. If the `process_priority` was in between it would force them into different bytes afaik. So what was once probably about 64 bytes is now probably in about 4.

lawnjelly commented on core/math/transform_interpolator.cpp:
```diff
@@ -0,0 +1,373 @@
+#include "core/engine.h"
+#include "core/math/transform.h"
+#include "core/print_string.h"
+#include "core/variant.h"
```
Correct! Done. :+1:

lawnjelly commented on doc/classes/SceneTree.xml:
```diff
@@ -215,6 +221,13 @@
 				To handle 'Go Back' button when this option is disabled, use [constant MainLoop.NOTIFICATION_WM_GO_BACK_REQUEST].
 			</description>
 		</method>
+		<method name="set_scene_tree_physics_interpolation_enabled">
```
Have done this, called `scene_tree_physics_interpolation`.

Wasn't sure about whether to have `scene_tree` as part of the name as it is part of `SceneTree`, so that kind of follows, but we are ending up with quite a few things called `physics_interpolation` and maybe something like this would disambiguate it.

Let me know if e.g. `physics_interpolation` is preferred, or something better.

akien-mga replied:
Yeah I think `scene_tree` is redundant when we are in the `SceneTree` class, both for the property name and the functions. It's implicit that all methods and properties of `SceneTree` impact the whole tree.

lawnjelly replied:
Renamed to `physics_interpolation` / `physics_interpolation_enabled`.

akien-mga reviewed (approved):
Looks good to me!

@reduz wanted a thorough review by @clayjohn, which could still be done post-merge - let's get this included in the next 3.5 beta to get broader testing.

lawnjelly commented:
> @reduz wanted a thorough review by @clayjohn, which could still be done post-merge - let's get this included in the next 3.5 beta to get broader testing.

Great. BTW I've also done preliminary docs for the interpolation and occluder polys, which might be good to link in the beta notes:

https://github.com/lawnjelly/Misc/blob/master/FTIDocs/FTI.md
https://github.com/lawnjelly/Misc/blob/master/OccluderDocs/OccluderShapePolygon.md

I'll be improving these and doing some screenshots etc as I get the chance.

akien-mga commented:
Thanks!

clayjohn commented:
Looks good to me too! My apologies for taking so long to review the full thing

mackatap commented:
I'm not sure if this should be submitted as a separate issue.  Enabling physics interpolation seems to double the speed scale on the second and subsequent emissions of CPUParticles. Halving the speed scale to .5 makes the particles act as they should.  The first emission seems to work as intended. I could upload a minimal project if needed, but it is just any CPUParticles.

Diddykonga commented:
> 

https://github.com/godotengine/godot/issues
Any issues with already merged features should be posted there. 
Please make sure to provide a minimal reproduction project, as it makes it significantly easier to find the issue.

lawnjelly commented:
There are some significant changes to CPUParticles in the PR. This would be good posted as an issue with an MRP, meanwhile I will try to replicate it. It could be related to something being done / advanced twice, as we have to deal with both previous and current frames when using interpolation.

marcinn commented:
Hi. 

v3.5.beta.custom_build (from origin/3.x branch). I'm noticing issues with interpolation of collision shapes. It looks like they doesn't react to `NOTIFICATION_RESET_PHYSICS_INTERPOLATION`, which should be received "by the node and all children recursively" (as docs says).

https://www.youtube.com/watch?v=hVUG-7baPw4

lawnjelly commented:
> v3.5.beta.custom_build (from origin/3.x branch). I'm noticing issues with interpolation of collision shapes. It looks like they doesn't react to `NOTIFICATION_RESET_PHYSICS_INTERPOLATION`, which should be received "by the node and all children recursively" (as docs says).

This is actually expected as far as I can see, nothing is technically going wrong. Collision shapes themselves are _not interpolated_ (they expressly should not be), they are not derived from `VisualInstance`, and outside the editor they are not even visible. So they do not receive the reset physics interpolation signal.

The wireframe in the editor for the debug collision shapes is purely for debugging, and is attached to these non-interpolating collision shapes. The wireframe does seem to currently be interpolated (which is probably just the default).

It does look a little strange when teleporting I will admit. I can have a look into whether it can be made to match the graphical sphere (if you can create an issue with an MRP), or maybe we can simply turn off the interpolation flag for the wireframe, that might be better for visualization. But it is not a nasty "bug" per se, just an interesting side effect (that's my excuse lol :grin: ).

marcinn commented:
I have issues with collision detection when phys interp is on. I have an area containing (0,0,0), and when I'm adding a player to the scene and immediately "teleporting" it to the desired position, the `body_entered` signal is emitting despite the player is moved (and drawn) outside the area.

Maybe I'm doing something wrong, so I would ask how to add a node to the scene into the desired place? I cannot set it's position before adding to the tree - Godot complains about this.

lawnjelly commented:
> I have issues with collision detection when phys interp is on. I have an area containing (0,0,0), and when I'm adding a player to the scene and immediately "teleporting" it to the desired position (outside the area), the `body_entered` signal is emitting despite the player is moved (and drawn) outside the area.
> 
> Maybe I'm doing something wrong, so I would ask how to add a node to the scene into the desired place? I cannot set it's position before adding to the tree - Godot complains about this.

This needs an issue and a minimum reproduction project, and I can look into it. :+1:
